File: acapella_app/lib/services/track_service.dart
import 'package:dio/dio.dart';
import '../models/track.dart';
import 'api_client.dart';

class TrackService {
  final Dio _dio = ApiClient().dio;

  Future<List<Track>> getAllTracks({int page = 1, int? albumId}) async {
    final query = {'page': page};
    if (albumId != null) query['album_id'] = albumId;

    final response = await _dio.get('/tracks', queryParameters: query);
    if (response.statusCode == 200) {
      final dataList = response.data['data']['data'] as List;
      return dataList.map((j) => Track.fromJson(j)).toList();
    }
    throw Exception('Failed to load tracks.');
  }

  Future<Track> getTrack(int id) async {
    final response = await _dio.get('/tracks/$id');
    if (response.statusCode == 200) {
      return Track.fromJson(response.data['data']);
    }
    throw Exception('Failed to load track.');
  }

  // Interaction feature
  Future<void> likeTrack(int id) async {
    await _dio.post('/tracks/$id/like');
  }

  Future<void> unlikeTrack(int id) async {
    await _dio.delete('/tracks/$id/like');
  }

  // Manager functions
  Future<Track> createTrack(Map<String, dynamic> data) async {
    final response = await _dio.post('/tracks', data: data);
    if (response.statusCode == 201) {
      return Track.fromJson(response.data['data']);
    }
    throw Exception('Failed to create track.');
  }

  Future<Track> updateTrack(int id, Map<String, dynamic> data) async {
    final response = await _dio.put('/tracks/$id', data: data);
    if (response.statusCode == 200) {
      return Track.fromJson(response.data['data']);
    }
    throw Exception('Failed to update track.');
  }

  Future<void> deleteTrack(int id) async {
    final response = await _dio.delete('/tracks/$id');
    if (response.statusCode != 200) {
      throw Exception('Failed to delete track.');
    }
  }
}

File: acapella_api_2/resources/js/Services/trackService.js
import apiService, { extractData } from './apiService';

/**
 * Track Service
 * Handles all track-related API calls
 */
const trackService = {
    /**
     * Get all tracks (public)
     */
    async getTracks(params = {}) {
        const response = await apiService.get('/tracks', { params });
        return extractData(response);
    },

    /**
     * Get a single track by ID (public)
     */
    async getTrack(trackId) {
        const response = await apiService.get(`/tracks/${trackId}`);
        return extractData(response);
    },

    /**
     * Create a new track (choir_manager, admin)
     */
    async createTrack(data) {
        const response = await apiService.post('/tracks', data);
        return extractData(response);
    },

    /**
     * Update a track (choir_manager, admin)
     */
    async updateTrack(trackId, data) {
        const response = await apiService.put(`/tracks/${trackId}`, data);
        return extractData(response);
    },

    /**
     * Delete a track (choir_manager, admin)
     */
    async deleteTrack(trackId) {
        const response = await apiService.delete(`/tracks/${trackId}`);
        return extractData(response);
    },

    /**
     * Like a track (authenticated)
     */
    async likeTrack(trackId) {
        const response = await apiService.post(`/tracks/${trackId}/like`);
        return extractData(response);
    },

    /**
     * Unlike a track (authenticated)
     */
    async unlikeTrack(trackId) {
        const response = await apiService.delete(`/tracks/${trackId}/like`);
        return extractData(response);
    },

    /**
     * Get related tracks (public)
     */
    async getRelatedTracks(trackId) {
        const response = await apiService.get(`/tracks/${trackId}/related`);
        return extractData(response);
    },

    /**
     * Get user's queue (authenticated)
     */
    async getQueue() {
        const response = await apiService.get('/queue');
        return extractData(response);
    },

    /**
     * Add track to queue (authenticated)
     */
    async addToQueue(trackId) {
        const response = await apiService.post(`/tracks/${trackId}/queue`);
        return extractData(response);
    },

    /**
     * Remove track from queue (authenticated)
     */
    async removeFromQueue(queueId) {
        const response = await apiService.delete(`/queue/${queueId}`);
        return extractData(response);
    },

    /**
     * Clear queue (authenticated)
     */
    async clearQueue() {
        const response = await apiService.delete('/queue');
        return extractData(response);
    },

    /**
     * Replace the queue with the given track order.
     * Used for shuffle and (future) drag-to-reorder.
     * @param {number[]} trackIds  ordered list of track ids
     */
    async reorderQueue(trackIds) {
        const response = await apiService.post('/queue/reorder', { track_ids: trackIds });
        return extractData(response);
    },
};

export default trackService;
